### PR TITLE
1180 refactor foxx authz

### PR DESCRIPTION
**Pull request by JoshuaSBrown**
# PR Description

# Tasks

* [x] - A description of the PR has been provided, and a diagram included if it is a new feature.
* [x] - Formatter has been run
* [x] - Labels have been assigned to the pr
* [x] - A reviwer has been added
* [x] - A user has been assigned to work on the pr
* [x] - If new feature a unit test has been added

## Summary by Sourcery

Refactor the authorization router by modularizing the code into separate modules for better maintainability and readability. Introduce new unit tests for the Record class and path handling to ensure robustness and correct error handling.

Enhancements:
- Refactor the authorization router to improve code readability and maintainability by modularizing the logic into separate modules.

Tests:
- Add new unit tests for the Record class and path handling to ensure correct functionality and error handling.

### CODE REVIEW

**Comment by JoshuaSBrown** [comment]
@sourcery-ai review

File: core/database/foxx/api/authz_router.js [hunk 1]
```diff
@@ -2,246 +1,250 @@
...
+          if (req.queryParams.act != "read" || !g_lib.hasPublicRead(data_id)) {
+            console.log(
+              "AUTHZ act: " + req.queryParams.act +
+                " client: " + client._id +
+                " path " + req.queryParams.file +
+                " FAILED"
+            );
+            throw g_lib.ERR_PERM_DENIED;
+          }
+        }
```
[the remark this answers is not in the document]
Comment (JoshuaSBrown) [reply core/database/foxx/api/authz_router.js:63]: I think this is preferntial, now I have to keep track of where the parentheses are being set.

File: core/database/foxx/api/authz_router.js [hunk 2]
```diff
@@ -1,247 +1,250 @@
...
+      // const path_components = pathModule.splitPOSIXPath(req.queryParams.file);
+      //
+      // Path components will be
+      // ["usr", "local", "bin"]
+
+      // Special case - allow unknown client to read a publicly accessible record
+      // if record exists and if it is a public record
+      if (!client) {
+        if (record.exists()) {
+          if (req.queryParams.act != "read" || !g_lib.hasPublicRead(data_id)) {
```
[the remark this answers is not in the document]
Comment (JoshuaSBrown) [reply core/database/foxx/api/authz_router.js:54]: Nice catch.

File: core/database/foxx/api/record.js [hunk 3]
```diff
@@ -0,0 +1,197 @@
+"use strict";
+
+const g_db = require("@arangodb").db;
+const g_lib = require("./support");
+const { errors } = require("@arangodb");
+
+/**
```
[the remark this answers is not in the document]
Comment (JoshuaSBrown) [reply core/database/foxx/api/record.js:7]: This is design preference.

**Comment by JoshuaSBrown** [comment]
#https://github.com/ORNL/DataFed/issues/1190

File: core/database/foxx/api/authz_router.js [hunk 4]
```diff
@@ -2,246 +1,250 @@
...
+        result = true,
+        id = g_lib.resolveID(req.queryParams.id, client),
+        ty = id[0];
+
+      if (id[1] != "/") {
+        throw [g_lib.ERR_INVALID_PARAM, "Invalid ID, " + req.queryParams.id];
+      }
+
+      if (ty == "p") {
+        var role = g_lib.getProjectRole(client._id, id);
```
[the remark this answers is not in the document]
Comment (JoshuaSBrown) [reply core/database/foxx/api/authz_router.js]: I did not touch this code except formatting,

File: core/database/foxx/api/authz_router.js [hunk 5]
```diff
@@ -2,246 +1,250 @@
...
+        if (role == g_lib.PROJ_NO_ROLE) {
+          // Non members have only VIEW permissions
+          if (perms != g_lib.PERM_RD_REC) result = false;
```
[the remark this answers is not in the document]
Comment (JoshuaSBrown) [reply core/database/foxx/api/authz_router.js]: I did not touch this code except formatting,

File: core/database/foxx/api/authz_router.js [hunk 6]
```diff
@@ -2,246 +1,250 @@
...
+        } else if (role == g_lib.PROJ_MEMBER) {
+          // Non members have only VIEW permissions
+          if ((perms & ~g_lib.PERM_MEMBER) != 0) result = false;
```
[the remark this answers is not in the document]
Comment (JoshuaSBrown) [reply core/database/foxx/api/authz_router.js]: I did not touch this code except formatting,

File: core/database/foxx/api/authz_router.js [hunk 7]
```diff
@@ -2,246 +1,250 @@
...
+  .queryParam("file", joi.string().required(), "Data file name")
+  .queryParam("act", joi.string().required(), "GridFTP action: 'lookup', 'chdir', 'read', 'write', 'create', 'delete'")
+  .summary("Checks authorization")
+  .description("Checks authorization");
+
+router
+  .get("/perm/check", function (req, res) {
+    try {
+      const client = g_lib.getUserFromClientID(req.queryParams.client);
+      var perms = req.queryParams.perms ? req.queryParams.perms : g_lib.PERM_ALL;
```
[the remark this answers is not in the document]
Comment (JoshuaSBrown) [reply core/database/foxx/api/authz_router.js]: I did not touch this code except formatting,

[the remark this answers is not in the document]
Comment (JoshuaSBrown) [reply core/database/foxx/api/authz_router.js]: I did not touch this code except formatting,

File: core/database/foxx/api/authz_router.js [hunk 8]
```diff
@@ -2,246 +1,250 @@
...
+        var role = g_lib.getProjectRole(client._id, id);
+        if (role == g_lib.PROJ_NO_ROLE) {
+          // Non members have only VIEW permissions
+          if (perms != g_lib.PERM_RD_REC) result = false;
+        } else if (role == g_lib.PROJ_MEMBER) {
+          // Non members have only VIEW permissions
+          if ((perms & ~g_lib.PERM_MEMBER) != 0) result = false;
+        } else if (role == g_lib.PROJ_MANAGER) {
+          // Managers have all but UPDATE
+          if ((perms & ~g_lib.PERM_MANAGER) != 0) result = false;
```
[the remark this answers is not in the document]
Comment (JoshuaSBrown) [reply core/database/foxx/api/authz_router.js]: I did not touch this code except formatting,

**Comment by JoshuaSBrown** [comment]
https://github.com/ORNL/DataFed/issues/1191

File: core/database/foxx/api/record.js [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+
+    // If there is a new repo we need to check the path there and use that
+    if (this.#loc.new_repo) {
+      // Below we get the allocation associated with data item by
+      // 1. Checking if the data item is in flight, is in the process
+      // of being moved to a new location or new owner and using that
+      // oweners id.
+      // 2. Using the loc.uid parameter if not inflight to get the owner
+      // id.
+      let new_alloc = g_db.alloc.firstExample({
```
Comment (AronPerez) [inline core/database/foxx/api/record.js:171]: If we aren't going to reassign this, we can use `const` to assign the variable

```suggestion
      const new_alloc = g_db.alloc.firstExample({
```

File: core/database/foxx/api/record.js [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+        this.#err_msg =
+          "Permission denied, '" +
+          this.#key +
+          "' is not part of an allocation '";
+        return false;
+      }
+
+      let stored_path = this._pathToRecord(new_alloc.path);
+
+      if (!this._comparePaths(stored_path, a_path)) { return false; }
```
Comment (AronPerez) [inline core/database/foxx/api/record.js:189]: Very nice 🧼

File: core/database/foxx/tests/record.test.js [hunk 11]
```diff
@@ -0,0 +1,341 @@
+"use strict";
+
+const chai = require("chai");
```
Comment (AronPerez) [inline core/database/foxx/tests/record.test.js:3]: Is there a reason we use Chai over Jest? 

I like Jest since it has builtin [sinon](https://github.com/sinonjs/sinon) functionality with [spyOn](https://jestjs.io/docs/jest-object#jestspyonobject-methodname) and [mockImplementation](https://jestjs.io/docs/mock-functions).

File: core/database/foxx/tests/record.test.js [hunk 12]
```diff
@@ -0,0 +1,341 @@
...
+    expect(record.key()).to.equal(valid_key);
+    expect(record.error()).to.be.null;
+    expect(record.errorMessage()).to.be.null;
+  });
+
+  it("unit_record: isManaged should initialize correctly, but show a record as not managed.", () => {
+    const valid_key = "1111";
+    const key_id = "d/1111";
+    const owner_id = "u/bob";
+    const repo_id = "repo/datafed-at-com";
```
Comment (AronPerez) [inline core/database/foxx/tests/record.test.js:66]: We can pull this common logic out at the top of the test suite since we use it across it.

File: core/database/foxx/tests/record.test.js [hunk 13]
```diff
@@ -0,0 +1,341 @@
...
+    // Create edges
+    g_db.loc.save({
+      _from: key_id,
+      _to: repo_id,
+      uid: owner_id,
+    });
+    g_db.alloc.save({
+      _from: owner_id,
+      _to: repo_id,
+    });
```
Comment (AronPerez) [inline core/database/foxx/tests/record.test.js:54]: If this setup is common we can pull it out as a function and pass in what we need ahead.

File: core/database/foxx/tests/record.test.js (hunk 11 above)
Comment (AronPerez) [reply core/database/foxx/tests/record.test.js:3]: Turns out it's cause ArangoDB comes pre-packaged with it, all good here

**Review by AronPerez — CHANGES_REQUESTED** [review]

File: core/database/foxx/tests/record.test.js (hunk 12 above)
Comment (JoshuaSBrown) [reply core/database/foxx/tests/record.test.js:66]: I have a better suggestion, I would prefer to make it unique for each test, that way if something goes wrong it is easier to track which test the log output is associated with.

File: core/database/foxx/api/support.js [hunk 14]
```diff
@@ -595,7 +595,21 @@ module.exports = (function() {
...
+      //   "name_first" : "bob",
+      //   "name_last" : "jones",
+      //   "is_admin" : true,
+      //   "max_coll" : 50,
+      //   "max_proj" : 10,
+      //   "max_sav_qry" : 20,
+      //   :
+      //   "email" : "bobjones@gmail.com"
+      // }
+      obj.getUserFromClientID = function(a_client_id) {
```
Comment (AronPerez) [inline core/database/foxx/api/support.js:612]: Was this supposed to be JSDoc format?

File: core/database/foxx/tests/authz.test.js [hunk 15]
```diff
@@ -0,0 +1,56 @@
+"use strict";
+
+const chai = require("chai");
+const expect = chai.expect;
+const authzModule = require("../api/authz"); // Replace with the actual file name
```
Comment (AronPerez) [inline core/database/foxx/tests/authz.test.js:5]: We can remove this comment

File: core/database/foxx/tests/authz.test.js [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+
+  beforeEach(() => {
+    g_db.d.truncate();
+    g_db.alloc.truncate();
+    g_db.loc.truncate();
+    g_db.repo.truncate();
+    g_db.u.truncate();
+  });
+
+  it("unit_authz: if admin should return true", () => {
```
Comment (AronPerez) [inline core/database/foxx/tests/authz.test.js:20]: Don't we want to test if not admin?

**Review by AronPerez — COMMENTED** [review]
Questions about testing and some comments left in

File: core/database/foxx/tests/authz.test.js (hunk 16 above)
Comment (JoshuaSBrown) [reply core/database/foxx/tests/authz.test.js:20]: This is addressed, in the next pr, see here, but good spot.

https://github.com/ORNL/DataFed/pull/1196/files#diff-29f2f796d64cb649168cf5d450247a757a3f3b2d0c252f8ad394c76ac48ee08c

File: core/database/foxx/api/support.js (hunk 14 above)
Comment (JoshuaSBrown) [reply core/database/foxx/api/support.js:612]: No, was just trying to be helpful, with additional information, I wasn't trying to follow a specific format yet. I propose a separate PR. To add JS formatting, I don't really want to touch this PR since it is passing the CI.

Comment (JoshuaSBrown) [reply core/database/foxx/api/support.js:612]: Addressed here:

https://github.com/ORNL/DataFed/pull/1196/files#diff-059ab4bef6420d0e2917fb00d6d95ff4ee621c6646e96ac1ea0fd5e9fa5c8633R598-R631

File: core/database/foxx/tests/authz.test.js (hunk 15 above)
Comment (JoshuaSBrown) [reply core/database/foxx/tests/authz.test.js:5]: https://github.com/ORNL/DataFed/pull/1196/files#diff-29f2f796d64cb649168cf5d450247a757a3f3b2d0c252f8ad394c76ac48ee08cL5

File: core/database/foxx/api/support.js [hunk 17]
```diff
@@ -1001,6 +1034,31 @@ module.exports = (function () {
         return false;
     };
 
+    /**
+     * @breif checks to make sure the client has admin permissions on an object
```
Comment (t-ramz) [inline core/database/foxx/api/support.js:1038]: Typo? `@brief`

File: core/database/tests/test_foxx.sh [hunk 18]
```diff
@@ -128,6 +128,7 @@ if ! command -v foxx > /dev/null 2>&1; then
 fi
 
 PATH_TO_PASSWD_FILE=${SOURCE}/database_temp.password
+echo "$local_DATAFED_DATABASE_PASSWORD" > "${PATH_TO_PASSWD_FILE}"
```
Comment (t-ramz) [inline core/database/tests/test_foxx.sh:131]: Did this cause any issues? I believe the file is made in the setup script, right?

File: core/database/foxx/tests/path.test.js
Comment (t-ramz) [inline core/database/foxx/tests/path.test.js:1]: Should there also be a test for relative paths?

File: core/database/foxx/api/authz_router.js [hunk 19]
```diff
@@ -5,6 +5,9 @@ const router = createRouter();
 const joi = require("joi");
 const g_db = require("@arangodb").db;
 const g_lib = require("./support");
+const pathModule = require("./posix_path"); // Replace with the actual file name
```
Comment (t-ramz) [inline core/database/foxx/api/authz_router.js:8]: I believe these comments can be removed.

File: core/database/foxx/api/authz_router.js [hunk 20]
```diff
@@ -22,44 +25,67 @@ router
...
+            //   "max_sav_qry" : 20,
+            //   :
+            //   "email" : "bobjones@gmail.com"
+            // }
             const client = g_lib.getUserFromClientID_noexcept(req.queryParams.client);
 
-            var idx = req.queryParams.file.lastIndexOf("/");
-            var data_key = req.queryParams.file.substr(idx + 1);
-            var data_id = "d/" + data_key;
+            // Will split a posix path into an array
```
Comment (t-ramz) [inline core/database/foxx/api/authz_router.js:44]: I think this comment is unnecessary because of the `splitPOSIXPath` jsdoc

File: core/database/foxx/api/authz_router.js [hunk 21]
```diff
@@ -70,85 +96,75 @@ router
...
+                            " client: " +
+                            client._id +
+                            " path " +
+                            req.queryParams.file +
+                            " FAILED",
+                    );
+                    throw [record.error(), record.errorMessage()];
+                }
+            } else {
+                // If the record does not exist then the path would noe be consistent.
```
Comment (t-ramz) [inline core/database/foxx/api/authz_router.js:131]: Typo? noe -> not ?

File: core/database/foxx/api/authz_router.js [hunk 22]
```diff
@@ -70,85 +96,75 @@ router
...
-            //var path = req.queryParams.file.substr( req.queryParams.file.indexOf("/",8));
-            var path = req.queryParams.file;
-            console.log("data_id is, ", data_id);
-            var loc = g_db.loc.firstExample({
-                _from: data_id,
-            });
-            console.log("Loc is:");
-            console.log(loc);
-            if (!loc) {
+            if (record.exists()) {
```
Comment (t-ramz) [inline core/database/foxx/api/authz_router.js:117]: There are a lot of checks for record existence. Is there a way to pull this check up a level or two in the nested conditionals?

File: core/database/foxx/api/record.js [hunk 23]
```diff
@@ -0,0 +1,200 @@
...
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * @brief Checks if the record exists in the database.
+     * @return {boolean} True if the record exists, otherwise false.
+     */
+    exists() {
```
Comment (t-ramz) [inline core/database/foxx/api/record.js:87]: Ok I see the `exists()` calls are inexpensive. That's good.

File: core/database/foxx/tests/authz.test.js [hunk 24]
```diff
@@ -0,0 +1,53 @@
...
+describe("Authz functions", () => {
+    beforeEach(() => {
+        g_db.d.truncate();
+        g_db.alloc.truncate();
+        g_db.loc.truncate();
+        g_db.repo.truncate();
+        g_db.u.truncate();
+    });
+
+    it("unit_authz: if admin should return true", () => {
```
Comment (t-ramz) [inline core/database/foxx/tests/authz.test.js:19]: Should we also have a test for not admin, but has write? And another for no admin, no write?

**Review by t-ramz — COMMENTED** [review]
I have a few questions and comments!

File: core/database/foxx/api/authz.js [hunk 25]
```diff
@@ -0,0 +1,59 @@
...
+     * }
+     *
+     * @param - the permission type that is being checked i.e.
+     *
+     * PERM_CREATE
+     * PERM_WR_DATA
+     * PERM_RD_DATA
+     **/
+    obj.isRecordActionAuthorized = function (a_client, a_data_key, a_perm) {
+        const data_id = "d/" + a_data_key;
```
[the remark this answers is not in the document]
Comment (JoshuaSBrown) [reply core/database/foxx/api/authz.js:44]: Great question, and yes you can. Thie is from the official documentation.

https://docs.arangodb.com/stable/concepts/data-structure/documents/

All documents contain special attributes at the top-level that start with an underscore, known as system attributes:

The document key is stored as a string in the _key attribute.
The document identifier is stored as a string in the _id attribute.
The document revision is stored as a string in the _rev attribute.
You can specify a value for the _key attribute when creating a document. The _id attribute is automatically set based on the collection and _key. The _id and _key values are immutable once the document has been created. The _rev value is maintained by ArangoDB automatically.

keys are unique to the collection, ids globally unique.

[the remark this answers is not in the document]
Comment (JoshuaSBrown) [reply core/database/foxx/api/authz.js:44]: I would say that if you don't conform to the pattern, it is likely an anti-pattern.

File: core/database/foxx/api/authz_router.js (hunk 22 above)
Comment (JoshuaSBrown) [reply core/database/foxx/api/authz_router.js:117]: Good catch, This is addressed in a follow up PR. see here. https://github.com/ORNL/DataFed/pull/1196/files#diff-2a4e90aa74276bb3ccc3605ae33282b2cfa475fd702f7239150b202e6a4ed152

File: core/database/foxx/tests/authz.test.js (hunk 24 above)
Comment (JoshuaSBrown) [reply core/database/foxx/tests/authz.test.js:19]: These are addressed in a follow up PR here: https://github.com/ORNL/DataFed/pull/1196/files#diff-29f2f796d64cb649168cf5d450247a757a3f3b2d0c252f8ad394c76ac48ee08c

File: core/database/tests/test_foxx.sh (hunk 18 above)
Comment (JoshuaSBrown) [reply core/database/tests/test_foxx.sh:131]: Good point. This shouldn't be in here.

File: core/database/foxx/tests/path.test.js
Comment (JoshuaSBrown) [reply core/database/foxx/tests/path.test.js:1]: I'm not sure there would be a benefit, all this function is doing is splitting on a delimiter. What edge case are you thinking could be problematic?

I can see this being a problem for code using the method if they are assuming a folder exists. 

i.e. 

dirC/dirA/../dirB/

Then the second folder would be dirC not '..', but that check isn't required to be in the splitPosixPath function. Unless you think it is confusing. All paths coming from gridFTP will be absolute paths.

File: core/database/foxx/api/authz.js (hunk 25 above)
[the remark this answers is not in the document]
Comment (t-ramz) [reply core/database/foxx/api/authz.js:44]: Ok yeah I see that in the docs. Cool to know.

File: core/database/foxx/api/authz.js [hunk 26]
```diff
@@ -0,0 +1,59 @@
...
+     *   "max_coll" : 50,
+     *   "max_proj" : 10,
+     *   "max_sav_qry" : 20,
+     *   :
+     *   "email" : "bobjones@gmail.com"
+     * }
+     *
+     * @param - the permission type that is being checked i.e.
+     *
+     * PERM_CREATE
```
[the remark this answers is not in the document]
Comment (JoshuaSBrown) [reply core/database/foxx/api/authz.js:39]: That's a good suggestion. The client is not currently defined in a single location as it is what is pulled out of the database.

**Review by AronPerez — APPROVED** [review]
